.. raw:: html

    <style> .navy {color:navy} </style>
	
.. role:: navy

.. raw:: html

    <style> .white {color:white} </style>

.. role:: white

.. raw:: html

    <style> .green {color:green} </style>
    <style> .blue {color:blue} </style>
    <style> .red {color:red} </style>

.. role:: green
.. role:: blue
.. role:: red


#############
Getting ready
#############

Install the SMAP package and download the tutorial data from Zenodo.

####################
Analysis of GBS data
####################

Activate your virtual environment ::

	source .venv/bin/activate

SMAP delineate
~~~~~~~~~~~~~~

Create and navigate to a new output directory to run **SMAP delineate** on the BAM files with mapped GBS reads of a set of **individuals**::

	mkdir -p download_dir/SMAP_delineate/03_output/ind
	cd download_dir/SMAP_delineate/03_output/ind/
	smap delineate download_dir/SMAP_delineate/02_input/ind/ -mapping_orientation ignore --processes 1 --plot all --plot_type pdf --min_stack_depth 2 --max_stack_depth 1500 --min_cluster_length 50 --max_cluster_length 300 --max_stack_number 20 --min_stack_depth_fraction 10 --min_cluster_depth 10 --max_cluster_depth 1500 --max_smap_number 20 --name 48_ind_GBS-PE 

Create and navigate to a new output directory to run **SMAP delineate** on the BAM files with mapped GBS reads of a set of **pool samples**::

	mkdir -p download_dir/SMAP_delineate/03_output/pools
	cd download_dir/SMAP_delineate/03_output/pools/
	smap delineate download_dir/SMAP_delineate/02_input/pools/ -mapping_orientation ignore --processes 1 --plot all --plot_type pdf --min_stack_depth 2 --max_stack_depth 1500 --min_cluster_length 50 --max_cluster_length 300 --max_stack_number 20 --min_stack_depth_fraction 5 --min_cluster_depth 30 --max_cluster_depth 1500 --max_smap_number 20 --name 16_pools_GBS-PE 


SMAP compare
~~~~~~~~~~~~

Create and navigate to a new output directory to run **SMAP compare** on the two BED files with MergedClusters generated by **SMAP delineate**::

	mkdir download_dir/SMAP_compare/03_output 
	cd download_dir/SMAP_compare/03_output
	smap compare download_dir/SMAP_delineate/03_output/ind/final_stack_positions_48_ind_GBS-PE_C0_SMAP20_CL50_300.bed download_dir/SMAP_delineate/03_output/pools/final_stack_positions_16_pools_GBS-PE_C0_SMAP20_CL50_300.bed 



SMAP haplotype-sites
~~~~~~~~~~~~~~~~~~~~

Create and navigate to a new output directory to run **SMAP haplotype-sites** on the BAM files with mapped GBS reads of a set of **individuals**, its BED file from **SMAP delineate**, and a VCF file with SNP calls (third-party software)::

	mkdir -p download_dir/SMAP_haplotype_sites/03_output/ind
	cd download_dir/SMAP_haplotype_sites/03_output/ind
	smap haplotype-sites download_dir/SMAP_haplotype_sites/02_input download_dir/SMAP_delineate/02_input/ind/final_stack_positions_48_ind_GBS-PE_C0_SMAP20_CL50_300.bed download_dir/SMAP_delineate/02_input/ind/48_ind_GBS-PE.vcf --out haplotypes_48_ind_GBS-PE -mapping_orientation ignore --discrete_calls dosage --frequency_interval_bounds diploid --dosage_filter 2 --plot all --plot_type pdf -partial include --min_distinct_haplotypes 2 --min_read_count 10 --min_haplotype_frequency 5 --processes 1

Create and navigate to a new output directory to run **SMAP haplotype-sites** on the BAM files with mapped GBS reads of a set of **pool samples**, its BED file from **SMAP delineate**, and a VCF file with SNP calls (third-party software)::

	mkdir -p download_dir/SMAP_haplotype_sites/03_output/pools
	cd download_dir/SMAP_delineate/03_output/pools/
	smap haplotype-sites download_dir/SMAP_haplotype_sites/02_input download_dir/SMAP_delineate/02_input/pools/final_stack_positions_16_pools_GBS-PE_C0_SMAP20_CL50_300.bed download_dir/SMAP_delineate/02_input/pools/48_ind_GBS-PE.vcf --out haplotypes_16_pools_GBS-PE -mapping_orientation ignore --plot all --plot_type pdf --mask_frequency 1 --undefined_representation "" -partial include --min_distinct_haplotypes 2 --min_read_count 30 --min_haplotype_frequency 5 --processes 1

Deactivate your virtual environment::

	deactivate

----

#######################
Analysis of HiPlex data
#######################


####################
Analysis of WGS data
####################


######################################
Analysis of CRISPR genome editing data
######################################

Activate your virtual environment::

	source .venv/bin/activate

Create and navigate to a new output directory to run **SMAP haplotype-window** on a set of FASTQ files with HiPlex reads, their mapped BAM files, its GFF file with border positions and a gRNA FASTA file from **SMAP design**:: 

	mkdir -p download_dir/SMAP_haplotype_window/03_output
	cd download_dir/SMAP_haplotype_window/03_output
	smap haplotype-window download_dir/SMAP_haplotype_window/02_input/reference.fasta download_dir/SMAP_haplotype_window/02_input/borders_gff download_dir/SMAP_haplotype_window/02_input/ download_dir/SMAP_haplotype_window/02_input/ --mask_frequency 2 --undefined_representation "" --min_read_count 30 --min_haplotype_frequency 5 --processes 1

Deactivate your virtual environment::

	deactivate

