.. SMAP documentation master file, created by
   sphinx-quickstart on Wed Aug  5 13:28:17 2020.
   You can adapt this file completely to your liking, but it should at least
   contain the root `toctree` directive.


.. _SMAPeffectindex:

SMAP effect-prediction
======================

| This is the manual for the component **SMAP effect-prediction** of the SMAP-package.
| **SMAP effect-prediction** was created to predict the effects of naturally occuring sequence variants or CRISPR-cas induced mutations in a given reference gene sequence on the encoded protein. 
| **SMAP effect-prediction** takes as input a genotyping mastertable generated by **SMAP haplotype-window** that lists the relative frequency of observed short haplotypes per locus per sample. 
| **SMAP effect-prediction** substitutes the reference sequence by the observed haplotype (taking the relative position of that haplotype in the reference gene and the relative positions of CDS into account), and predicts the resulting encoded protein for each haplotype detected. 
| **SMAP effect-prediction** then aligns the mutated protein with the reference protein, and scores the %identity (based on alignment completeness and amino-acid similarity) to the reference protein, in order to quantify the degree of "loss of function" as a result of the detected haplotype (compared to the reference protein. 
| **SMAP effect-prediction** then discretizes the strength of the loss of function effects per haplotype, according to user-defines effect-strength intervals, into the classes "no effect", "intermediate", "strong" effect per haplotype.  
| **SMAP effect-prediction** then aggregates the relative frequency of haplotypes with the same effect class per sample, (e.g. thus summing up the total frequency of strong effect haplotypes). 
| **SMAP effect-prediction** then transforms the quantitative frequencies per effect class per locus per sample into discrete dosage genotype calls per locus per sample (e.g. homozygous strong effect, heterozygous strong effect/no effect, homozygous no effect). 
| **SMAP effect-prediction** thus creates the input files required for downstream analysis such as phenotpe-genotype associations. 

The scheme below shows how the subsequent steps of **SMAP effect-prediction** filter, annotate, aggregate, and transform haplotype data and finally create discretized "effect-based" genotype call table.

			 .. image:: images/example_data/102.png

.. toctree::
   :maxdepth: 2
   :caption: Contents:
   
   effect_scope_quickstart
   effect_feature_definition
   effect_HIW
   effect_example_data
   effect_faq
   effect_code

